blog/content/post/2022-06-29-xen-debugging-and-development-environment.md

```diff
@@ -0,0 +1,170 @@
+
+More to that all of the methods can be applied through frameworks.
+Debian rootfs can be built using [isar](https://github.com/ilbers/isar) and
+there is always a way to narrow everything to OpenEmbedded/Yocto meta layer
+which should build only what we need. The last option is good for production,
+but development may be hard in a limited environment that Yocto produces by
+default.
+
+## Xen dockerized building environment
+
+I chose the second option for that purpose. I have prepared a Docker
+container with all required packages
+([3mdeb/xen-docker](https://github.com/3mdeb/xen-docker)):
+
+```bash
+$ git clone git@github.com:3mdeb/xen-docker.git
+$ cd xen-docker
+$ docker build -t 3mdeb/xen-docker .
+```
+
+I cloned the Xen source and I run the docker container.
+
+```bash
+$ git clone git@github.com:xen-project/xen.git
+$ cd xen
+$ docker run --rm -it -v $PWD:/home/xen -w /home/xen 3mdeb/xen-docker /bin/bash
+```
+Then I built the Xen package:
+
+```
+(docker-container)$ git checkout <version>
+(docker-container)$ ./configure --enable-githttp --enable-systemd
+# there is time now to customize .config
+(docker-container)$ make debball
+```
+
+> Note: `--enable-systemd` requires a `libsystemd-dev` package to be installed
+> in the container.
+
+Build result will be placed in `$XEN_SRC_DIR/dist` as
+`xen-upstream-<version>.deb`. For Debian-based systems, it is easy to install
+it with `dpkg` or `apt`. The package contains all necessary components for
+the host OS along with the Xen kernel.
+
+> Note that the host OS still will require Dom0 Kernel for Xen
+
+To update the pxe-server with new Xen image, install the
+`xen-upstream-<version>.deb` in rootfs which hosts the VMs and copy the Xen
+kernel from `/boot/xen-<version>.gz` to tftpboot/httpboot directory (gunzip the
+kernel first).
+
+After whole this effort, I was able to boot my freshly built Xen on apu2c4 with
+Debian host and Debian guest OS. Now that I have prepared the developing
+procedure I can start narrowing down all the issues.
+
+# Infrastructure
+
+Nature of Embedded Systems Consulting company forced us to build reliable
+infrastructure for kernels and rootfs building and deployment. It took quite
+a lot of time since there are a lot of options, but not much was working
+according to our specs.
+
+Our virtualization platform of choice was Proxmox we had to stick to that since
+the transition was not an option. We started to put together our stack:
+
+* Proxmox as a virtualization platform
+* docker-machine as hosts (VMs) management tool
+* [docker-machine-driver-proxmox-ve](https://github.com/lnxbil/docker-machine-driver-proxmox-ve)
+  driver required for docker-machine so it can communicate with Proxmox
+* [RancherOS](https://rancher.com/rancher-os/) as our container OS
+* [isar](https://github.com/ilbers/isar) as rootfs building framework
+* [meta-virtualization](https://git.yoctoproject.org/cgit/cgit.cgi/meta-virtualization/) as a framework for building Xen from source
```

Comment:
@Asiderr this is extremely outdated. Rancher OS is dead now. Also we are no longer Embedded System Consulting :)

Reply:
Sure, I will fix this. Should I change this stack to the apu/debian based stack?

Reply:
Or should I just cut out this last paragraph? It seems to be not relevant, adds nothing the the topic IMO.

Reply:
@Asiderr let's think what is the purpose of this article?

- We present containerized environment to build Xen
- Why we need custom Xen? - we can build products based (e.g. [lpnRouter](https://www.youtube.com/watch?v=0gPXXlWc1bw) on that or support existing projects (e.g. [XCP-ng](https://xcp-ng.com/), although I'm not sure if it is possible since they rely on XenServer from Citrix, but to some extent there is relation here, there is also XCP-ng and Protectli relation, which may make sense from Dasharo security features, [OpenXT](https://openxt.org/). There was also some customer demand related to edge computing. We test Xen i 3mdeb infrastructure, if this procedure works TAT should be able to include it in validation - according to [this](https://docs.google.com/spreadsheets/d/1_uRhVo9eYeZONnelymonYp444zYHT_Q_qmJEJ8_XqJc/edit?usp=sharing) our test rely on Xen 4.8 :scream: 
- Custom Xen needs CI, which needs scalable (dockerized/containerized) build environment

I expect to remove infrastructure paragraph or improve it to explain value this article can provide to community, our partners and 3mdeb.

Reply:
Imho, from a low-level perspective, the main purpose of this article is to show users/developers the easiest way to build custom Xen. Therefore, I will remove that last paragraph and point out possible use cases, highlighting our previous work and experience on this topic.

Reply:
@pietrushnic Paragraph is changed.